### PR TITLE
feat: add PARSE_DATE and FORMAT_DATE functions

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
Enables DATE data for UDFs and adds the PARSE_DATE and FORMAT_DATE functions + docs.

### Testing done 
QTT + unit tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 1]
```diff
@@ -1199,6 +1203,27 @@ TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, fo
 "America/Los_Angeles", "PDT", "Europe/London". For more information on timestamp formats, see
 [DateTimeFormatter](https://cnfl.io/java-dtf).
 
+### `FORMAT_DATE`
+
+```sql
+FORMAT_DATE(timestamp, 'yyyy-MM-dd')
+```
+
+Converts DATE value into a string representing the date in the given format.
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1212]:
```suggestion
Converts a DATE value into a string that represents the date in the given format.
```

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 2]
```diff
@@ -1199,6 +1203,27 @@ TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, fo
...
+Single quotes in the timestamp format can be escaped with two successive single
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1213]:
```suggestion
You can escape single-quote characters in the timestamp format by using two successive single
```

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 3]
```diff
@@ -1199,6 +1203,27 @@ TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, fo
...
+quotes, `''`, for example: `'yyyy-MM-dd''T'''`.
+
+### `PARSE_DATE`
+
+```sql
+PARSE_DATE(col1, 'yyyy-MM-dd')
+```
+
+Converts a string representation of a date in the
+given format into a DATE value. Single quotes in the timestamp
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1223]:
```suggestion
given format into a DATE value. You can escape single-quote characters in the timestamp
```

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 4]
```diff
@@ -1199,6 +1203,27 @@ TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, fo
...
+format can be escaped with two successive single
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1224]: as above

**Review by JimGalasyn — APPROVED** [review]
LGTM, with a couple of suggestions. Also needs to be cherry-[picked to the `0.20.x-ksqldb` branch.

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 5]
```diff
@@ -1199,6 +1203,27 @@ TIMEZONE is an optional parameter and it is a `java.util.TimeZone` ID format, fo
 "America/Los_Angeles", "PDT", "Europe/London". For more information on timestamp formats, see
 [DateTimeFormatter](https://cnfl.io/java-dtf).
 
+### `FORMAT_DATE`
+
+```sql
+FORMAT_DATE(timestamp, 'yyyy-MM-dd')
```
Comment (spena) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1209]: Shouldn't this be `FORMAT_DATE(date, ...)`

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatDate.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          description = "The date to convert") final Date date,
+      @UdfParameter(
+          description = "The format pattern should be in the format expected by"
+              + " java.time.format.DateTimeFormatter.") final String formatPattern) {
+    if (date == null) {
+      return null;
+    }
+    try {
+      final DateTimeFormatter formatter = formatters.get(formatPattern);
+      return LocalDate.ofEpochDay(TimeUnit.MILLISECONDS.toDays(date.getTime())).format(formatter);
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatDate.java:62]: I wonder what if `formatPattern` contains time characters? Would those be part of the final String?

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/FormatDateTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        });
+  }
+
+  @Test
+  public void shouldSupportEmbeddedChars() {
+    // When:
+    final Object result = udf.formatDate(Date.valueOf("2014-11-09"), "yyyy-dd-MM'Fred'");
+
+    // Then:
+    assertThat(result, is("2014-09-11Fred"));
```
Comment (spena) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/FormatDateTest.java:67]: I'm not sure about this. Should we support embedded chars in the final date string?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatDate.java (hunk 6 above)
Comment (jzaralim) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/FormatDate.java:62]: Added a test for this.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/FormatDateTest.java (hunk 7 above)
Comment (jzaralim) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/FormatDateTest.java:78]: It was supported in the old DateToString function: https://github.com/confluentinc/ksql/blob/cd1a9880aa39e0fb9b4e84d9f713235c8569b104/ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/DateToStringTest.java#L61

Comment (spena) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/datetime/FormatDateTest.java:78]: Just tested Mysql and it supports this too. Perhaps it is allowed in order to write the format you want, i.e. dates `yyyy/MM/dd` or `yyyy-MM-dd` or `Today is Year = yyyy Month = MM Day = dd`

**Review by spena — APPROVED** [review]
LGTM

**Pull request merged**